README
PDL::Graphics::PLplot::Map:

A PDL interface to the GMT (Generic Mapping Tools).  This allows one to access
a detailed coastline, river, lake and national boundary database covering the
entire world.  

This package includes a function, 'fetch', which will give raw lon/lat PDLs from
the GMT database.  It also includes the PDL::Graphics::PLplot compatible routines:

worldmap   -- Draw a world map using PGPLOT
map_plot   -- Add points or lines to an existing world map, including map projection

It currently supports two projects:  

-- Linear 
-- Azimuthal Equidistant

But others can easily be added.

The following are required for installation:

-- perl 5.5 or later
-- PDL v2.1 or later with bad value support compiled in (Set WITH_BADVAL => 1 in perldl.conf)
-- the PLplot plotting package (http://plplot.sourceforge.net/)
-- the PLplot perl interface PDL::Graphics::PLplot, see CPAN.
-- the netCDF library (see http://www.unidata.ucar.edu/packages/netcdf/index.html)

INSTALLATION:

Installation should be the normal:

perl Makefile.PL
make
make test

(as root)
make install

If the netcdf include or library files cannot be found, try:
   export NETCDF_LIBDIR='/ops/tools/lib'
   export NETCDF_INCDIR='/ops/tools/include'
   perl Makefile.PL
   ...

Best of luck!

Doug Hunt
dhunt@ucar.edu








Changes
Revision history for PDL::Graphics::Map:

0.01  07/27/2000
	- original version

0.06  5/24/2005
        - silly little fix to handle new bad value behavior in PDL 2.4.2

0.07  8/01/2005
        - Added  #include "XSUB.h" to pscoast.c.  Apparently newer perls (5.8.7)
          need this to find 'my_perl' symbol.
0.09  7/24/2008
        - Fix river plotting
0.10  10/15/2008
        - Add better prerequisite handling if PDL is not installed






test.pl
# Before `make install' is performed this script should be runnable with
# `make test'. After `make install' it should work as `perl test.pl'
use lib qw(./blib/lib ./blib/arch);
######################### We start with some black magic to print on failure.

# Change 1..1 below to 1..last_test_to_print .
# (It may become useful if the test is moved to ./t subdirectory.)

BEGIN { $| = 1; print "1..13\n"; }
END {print "not ok 1\n" unless $loaded;}
use PDL;
use PDL::Graphics::PLplot::Map;
$loaded = 1;
print "ok 1\n";

######################### End of black magic.

# Insert your test code below (better if it prints "ok 13"
# (correspondingly "not ok 13") depending on the success of chunk 13
# of the test code):

sub ok ($$) {
  my $bool = shift;
  my $n    = shift;
  if ($bool) {
    print "ok $n\n";
  } else {    print "not ok $n\n";
  }
}


my ($lon, $lat) = PDL::Graphics::PLplot::Map::fetch(RESOLUTION => 'crude');
my $pl = PDL::Graphics::PLplot->new (DEV => "png", FILE => "test2.png");
$pl->xyplot($lon, $lat);
$pl->close;
ok -s "test2.png" > 0, 2;

my $lat100 = pdl [79.15938, 79.621118,   79.3608, 80.058747,  78.90547, 79.788663, 79.263752, 79.579919, 79.877165, 79.546654, 78.204471, 78.850538, 78.457771, 78.716258, 78.429389, 78.056764, 77.716182, 77.929198, 77.653315, 77.459525, 77.524224, 76.999924, 76.999924,  76.59617,  78.42237, 78.548714,-999, 80.429236, 80.505531, 80.136263, 80.125887, 79.711147, 79.749905,-999, 78.526741, 78.206912, 78.903639, 78.527352, 78.526741,-999,70, 70.172427, 70.069581, 70.397345, 70.884718, 70.426947, 70.970169, 71.131304, 70.463264, 70.358282, 70.968032, 70.065309, 70.874037, 70.992142,70,-999, 79.749905, 79.802396, 79.452964, 79.181353, 79.943389,  79.94461, 80.346227, 80.103609, 80.504616,  79.97116, 80.429236,-999, 77.623713, 78.253605, 78.113222, 77.438773, 77.614557, 77.237049, 77.623713, 77.623713,-999, 78.204471, 78.612802, 78.480049, 78.205692, 78.204471,-999, 70.970474, 71.097429, 70.970474, 70.970474,-999,  70.33875, 70.621347,  70.33875,  70.33875,-999, 78.838636, 78.907607, 78.838636, 78.838636,-999, 70.266728, 70.334173];

my $tol = 1e-4;
ok ((all approx $lat100, $lat->slice("1:100"), $tol), 3);

my @bounds = (-180, 0, -90, 0);
my ($lon, $lat) = PDL::Graphics::PLplot::Map::fetch(RESOLUTION => 'low',
                                                      BOX        => \@bounds,
                                                      BOUNDARIES => [1],
                                                      COASTS     => 1);

my $pl = PDL::Graphics::PLplot->new (DEV => "png", FILE => "test4.png");
$pl->xyplot($lon, $lat);
$pl->close;
ok -s "test4.png" > 0, 4;

my @bounds = (0, 180, 0, 90);
my ($lon, $lat) = PDL::Graphics::PLplot::Map::fetch(RESOLUTION => 'low',
                                                      BOX        => \@bounds,
                                                      BOUNDARIES => [1],
                                                      COASTS     => 1);

my $pl = PDL::Graphics::PLplot->new (DEV => "png", FILE => "test6.png");
$pl->xyplot($lon, $lat);
$pl->close;
ok -s "test6.png" > 0, 6;

my $lat100 = pdl [80.40238, 80.505379, 80.204166,  80.33341, 80.136263,   80.0795, 80.202029, 80.125887, 80.020447, 80.043336,80,-99999, 80.282597, 80.233005, 80.299992, 80.363012, 80.290837, 80.282597,-99999, 80.536202, 80.536202,  80.46952, 80.536049,-99999,80, 80.058747,80,-99999,80, 80.009613,80,-99999,80, 80.150759,  80.15198,  80.27039, 80.241703, 80.346227, 80.243839, 80.314183, 80.347143, 80.269169, 80.278782, 80.121614,  80.19379, 80.103761, 80.420844, 80.504616, 80.311284, 80.416571, 80.396582,80,-99999,80, 80.110018, 80.110018, 80.260777, 80.196231, 80.404974, 80.384527,  80.40238,-99999, 80.128328, 80.135805, 80.141604, 80.128328,-99999, 80.649882, 80.615854, 80.654154, 80.649882,-99999, 80.384222, 80.443732, 80.389258, 80.384222,-99999, 80.668193,  80.67155, 80.674907, 80.668193,-99999, 80.536049, 80.536202, 80.536202,-99999, 80.077058, 80.220035, 80.085756, 80.076295, 80.077058,-99999, 80.132906, 80.171664, 80.132906, 80.132906,-99999, 80.923476, 80.780804, 80.674907];


ok ((all approx $lat100, $lat->slice("1:100"), $tol), 7);

#---------------------
# Start of plots
#---------------------

my $pl = PDL::Graphics::PLplot::Map->new (DEV => "png", FILE => "test8.png");
$pl->worldmap;
$pl->close;
ok -s "test8.png" > 0, 8;

my $pl = PDL::Graphics::PLplot::Map->new (DEV => "png", FILE => "test9.png");
$pl->worldmap (PROJECTION => 'AZEQDIST', # Azimuthal Equidistant projection
               CENTER  => [-170, 70],    # map centered at 170 deg west lon, 70 deg north lat
               RADIUS  => 3000);         # 3000 kilometer minimum radius
$pl->close;
ok -s "test9.png" > 0, 9;

# equatorial case
my $pl = PDL::Graphics::PLplot::Map->new (DEV => "png", FILE => "test10.png");
$pl->worldmap (PROJECTION => 'AZEQDIST', # Azimuthal Equidistant projection
               CENTER  => [-10,  20],    # map centered at 170 deg west lon, 70 deg north lat
               RADIUS  => 3000,          # 3000 kilometer minimum radius
	       RESOLUTION => 'low',
	       LONGRID => 10,            # longitude grid lines every 10 degrees
	       LATGRID => 10);           # latitude grid lines every 10 degrees

# plot projection center point
$pl->map_plot (pdl([-10]), pdl([20]), PLOTTYPE => 'POINTS',
                                      PROJECTION => 'AZEQDIST',
                                      CENTER  => [-10,  20], COLOR => 'GREEN');
$pl->close;
ok -s "test10.png" > 0, 10;

# polar case near date line
my $pl = PDL::Graphics::PLplot::Map->new (DEV => "png", FILE => "test11.png");
$pl->worldmap (PROJECTION => 'AZEQDIST', # Azimuthal Equidistant projection
               CENTER  => [-170,  80],   # map centered at 170 deg west lon, 80 deg north lat
               RADIUS  => 3000,          # 3000 kilometer minimum radius
	       RESOLUTION => 'low',
	       LONGRID => 20,            # longitude grid lines every 10 degrees
	       LATGRID => 10);           # latitude grid lines every 10 degrees

$pl->map_plot (pdl(-170, -160), pdl(80, 90), PLOTTYPE => 'LINE', 
                                             PROJECTION => 'AZEQDIST', 
                                             CENTER  => [-170,  80], COLOR => 'PINK');

$pl->close;
ok -s "test11.png" > 0, 11;


# lat lines messed up
my $pl = PDL::Graphics::PLplot::Map->new (DEV => "png", FILE => "test12.png");
$pl->worldmap (PROJECTION => 'AZEQDIST', # Azimuthal Equidistant projection
               CENTER  => [0, 90],
               RADIUS  => 5000,          # 3000 kilometer minimum radius
	       RESOLUTION => 'low',
	       LATGRID => 15);           # latitude grid lines every 10 degrees

$pl->close;
ok -s "test12.png" > 0, 12;

# lon lines messed up
my $pl = PDL::Graphics::PLplot::Map->new (DEV => "png", FILE => "test13.png");
$pl->worldmap (PROJECTION => 'AZEQDIST', # Azimuthal Equidistant projection
               CENTER  => [0, 90],
               RADIUS  => 5000,          # 3000 kilometer minimum radius
	       RESOLUTION => 'low',
	       LONGRID => 30);           # latitude grid lines every 10 degrees

$pl->close;
ok -s "test13.png" > 0, 13;














Makefile.PL
# Makefile.PL for a package defined by PP code.

use ExtUtils::MakeMaker;
use Config;

eval {
  require PDL::Core::Dev;            # Pick up development utilities
  PDL::Core::Dev->import;
  require PDL::Bad;                  # check for bad value support
};
my $pdl_installed = ($@) ? 0 : 1;

# -- Add new subroutines here! --

my @src = qw(pscoast.c gmt_init.c gmt_map.c gmt_shore.c gmt_support.c);
my @obj = @src;
map {s/\.[fc]/\.o/;} @obj; # swap .f, .c for .o

my $install = $Config{'installsitearch'};

# 
## Search for netcdf library and include file
#

foreach my $libdir (
		    $ENV{NETCDF_LIBDIR}, 
		    '/usr/local/netcdf/lib',
		    '/usr/local/lib',
		    '/opt/local/lib',
		    '/usr/lib',
		    '/opt/lib',
		                  # Add new library paths here!!
		    ) {

  if (-e "$libdir/libnetcdf.so") {
    $netcdf_lib_path = $libdir;
    print "Found libnetcdf.so at $libdir/libnetcdf.so\n";
    last;
  }

  if (-e "$libdir/libnetcdf.a") {
    $netcdf_lib_path = $libdir;
    print "Found libnetcdf.a at $libdir/libnetcdf.a\n";
    last;
  }

}

die "Cannot find netcdf library, libnetcdf.so or libnetcdf.a.  
Please install NetCDF or export NETCDF_LIBDIR with the correct library path" 
  unless defined ($netcdf_lib_path);

foreach my $incdir (
		    $ENV{NETCDF_INCDIR}, 
		    '/usr/local/netcdf/include',
		    '/usr/local/include',
		    '/opt/local/include',
		    '/usr/include',
		    '/opt/include',
		                  # Add new header paths here!!
		    ) {

  if (-e "$incdir/netcdf.h") {
    $netcdf_include_path = $incdir;
    print "Found netcdf.h at $incdir/netcdf.h\n";
    last;
  }
}

die "Cannot find netcdf header file, netcdf.h.
Please install NetCDF or export NETCDF_INCDIR with the correct include" 
  unless defined ($netcdf_include_path);

# Make sure bad value support is compiled in.
die "PDL::Graphics::PLplot::Map *requires* bad value support (Set WITH_BADVAL => 1 in perldl.conf)"
  if (!$PDL::Bad::Status && $pdl_installed);

# Check if compiled under gcc/Linux.  In which case, define bool for the compiler
$define_bool = '';
if ($Config{'osname'} =~ linux) {
  $define_bool = '-Dbool=int';
  print "Defining bool=int (linux seems to need this)\n";
}

my %pmfiles;
if ($pdl_installed) {
  %pmfiles = map { $_ => "\$(INST_LIBDIR)/Map/$_" } glob ("binned*cdf");
  $pmfiles{'Map.pm'} = '$(INST_LIBDIR)/Map.pm';
}

#print "pmfiles = \n";
#for (keys %pmfiles) { print "$_ => $pmfiles{$_}\n"; }

my $inc = '';
if ($pdl_installed) {
  $inc = &PDL_INCLUDE()." -I$netcdf_include_path -I./include";
}

$package = ["map.pd",Map,PDL::Graphics::PLplot::Map];
my $postamble = $pdl_installed ? pdlpp_postamble($package) : '';
WriteMakefile(
	      'NAME'  	     => 'PDL::Graphics::PLplot::Map',
	      'CCFLAGS'      => "$define_bool",
	      'PREREQ_PM'    => { PDL => 0 },
	      'PREREQ_PRINT' => 1,
              'DEFINE'       => 
	      "-D_SVID_SOURCE -DGMT_DEFAULT_PATH=\\\".\\\" -DGMT_INSTALL_PATH=\\\"$install/PDL/Graphics/PLplot/Map\\\"",  
	      'VERSION_FROM' => 'map.pd',
	      'OBJECT'       => 'Map.o ' . join (" ", @obj),
	      'PM'           => \%pmfiles,
#             'OPTIMIZE'     => '-g',
	      'INC'          => $inc,
	      'LIBS'         => ["-L$netcdf_lib_path -lnetcdf"],  
	      'clean'        => {'FILES'  => 
				   'Map.pm Map.xs Map.o Map.c pscoast.o testmap.png'},
	      'dist'         => { COMPRESS => 'gzip', SUFFIX => 'gz' },	
	      );
sub MY::postamble { $postamble; }












pscoast.c
 /*--------------------------------------------------------------------
 *    THE GMT-system:	@(#)pscoast.c	3.83  03/20/00
 *
 *	Copyright (c) 1991-2000 by P. Wessel and W. H. F. Smith
 *	See COPYING file for copying and redistribution conditions.
 *
 *	This program is free software; you can redistribute it and/or modify
 *	it under the terms of the GNU General Public License as published by
 *	the Free Software Foundation; version 2 of the License.
 *
 *	This program is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 *
 *	Contact info: www.soest.hawaii.edu/gmt
 *
 *--------------------------------------------------------------------*/
/*
 * pscoast (the expurgated version c/o Doug Hunt) can only dump binary
 * map data.  Three
 * data sets are considered: shores, borders, and rivers.  Each of these
 * data sets come in 5 different resolutions.  The lower resolution files
 * are derived from the full resolution data using the Douglas-Peucker (DP)
 * line reduction algorithm.  By giving a tolerance in km, the algorithm
 * will remove points that do not depart more than the tolerance from the
 * straight line that would connect the points if the point in question was
 * removed.  The resolutions are:
 *
 * full		- The complete World Vector Shoreline + CIA data base
 * high		- DP reduced with tolerance = 0.2 km
 * intermediate	- DP reduced with tolerance = 1 km
 * low		- DP reduced with tolerance = 5 km
 * crude	- DP reduced with tolerance = 25 km
 *
 * Author:	Paul Wessel
 * Date:	24-MAR-1984
 * Version:	3.1	Added clipping and dumping as options
 *
 * Excerpted by Doug Hunt to form the core of the PDL::Graphics::Map
 * interface to the GMT maps.  7/27/2000
 *
 */

#include "EXTERN.h"   /* std perl include */
#include "perl.h"     /* std perl include */
#include "XSUB.h"     /* std perl include */
#include "gmt.h"

#ifdef DEBUG
int debug_bin = -1;
#endif

struct GMT_SHORE c;
struct GMT_BR b, r;

char *shore_resolution[5] = {"full", "high", "intermediate", "low", "crude"};

void pscoast (double west, double east, double south, double north, char res, int rlevels[N_RLEVELS], int blevels[N_BLEVELS], int draw_coast, SV *lon, SV *lat)
{
	int i, n, np, ind, bin, anti_bin = -1, level, flag;
	int level_to_be_painted, max_level = MAX_LEVEL, direction, np_new, k, last_k;
	int n_blevels = 0, n_rlevels = 0, bin_trouble;
	int start_direction, stop_direction, lowest_level = 1, min_level = 0, n_pieces = 0;
	
	BOOLEAN	error = FALSE, set_lake_fill = FALSE, draw_river = FALSE, shift = FALSE, need_coast_base, recursive;
	BOOLEAN	fill_ocean = FALSE, fill_land = FALSE, possibly_donut_hell = FALSE;
	BOOLEAN clobber_background, draw_border = FALSE, paint_polygons = FALSE, donut, fill_in_use = FALSE;
	BOOLEAN draw_scale = FALSE, gave_xy = FALSE, fancy = FALSE, donut_hell = FALSE, world_map_save;
	BOOLEAN clip_ocean = FALSE, clip_land = FALSE, end_of_clip = FALSE, clipping = FALSE;
	
	double	bin_x[5], bin_y[5];
	double west_border, east_border, anti_lon, anti_lat, edge = 720.0, left, right;
	double *xtmp, *ytmp, min_area = 0.0, x0, y0, scale_lat, length, step = 0.01;
	double anti_x, anti_y, x_0, y_0, x_c, y_c, dist, out[2], mylon;
	
        char *myptr;
        int len, base, latlonlen = 0;

	char key[5], *string, measure = '\0', comment[32];

	struct POL *p;

	for (i=0;i<N_BLEVELS;i++) if (blevels[i]) n_blevels++;
	for (i=0;i<N_RLEVELS;i++) if (rlevels[i]) n_rlevels++;
	if (n_rlevels) draw_river  = TRUE;
	if (n_blevels) draw_border = TRUE;

	my_GMT_begin (); 

	base = GMT_set_resolution (&res, 'D');

	need_coast_base = (draw_coast);

	if (east > 360.0) {
		west -= 360.0;
		east -= 360.0;
	}
	
	GMT_map_getproject ("x1d");

	GMT_map_setup (west, east, south, north);

	world_map_save = GMT_world_map;
	
	if (need_coast_base && GMT_init_shore (res, &c, project_info.w, project_info.e, project_info.s, project_info.n))  {
		fprintf (stderr, "%s: %s resolution shoreline data base not installed\n", GMT_program, shore_resolution[base]);
		need_coast_base = FALSE;
	}
	
	if (draw_border && GMT_init_br ('b', res, &b, project_info.w, project_info.e, project_info.s, project_info.n)) {
		fprintf (stderr, "%s: %s resolution political boundary data base not installed\n", GMT_program, shore_resolution[base]);
		draw_border = FALSE;
	}
	
	if (draw_river && GMT_init_br ('r', res, &r, project_info.w, project_info.e, project_info.s, project_info.n)) {
		fprintf (stderr, "%s: %s resolution river data base not installed\n", GMT_program, shore_resolution[base]);
		draw_river = FALSE;
	}
	
	if (! (need_coast_base || draw_border || draw_river || end_of_clip)) {
		fprintf (stderr, "%s: No databases available - aborts\n", GMT_program);
		exit (EXIT_FAILURE);
	}

	start_direction = -1;
	stop_direction = 1;
	lowest_level = (fill_ocean) ? 0 : 1;

	if ((360.0 - fabs (project_info.e - project_info.w) ) < c.bsize)
		GMT_world_map = TRUE;
	
	/* west_border = project_info.w;	east_border = project_info.e; */

	/* Temporarily shift boundaries */

	/*
	if (project_info.w < 0.0 && project_info.e <= 0.0) {	
		project_info.w += 360.0;
		project_info.e += 360.0;
		if (project_info.central_meridian < 0.0) project_info.central_meridian += 360.0;
	}
	*/

	/* west_border = project_info.w;	east_border = project_info.e; */
	west_border = floor (project_info.w / c.bsize) * c.bsize;
	east_border = ceil (project_info.e / c.bsize) * c.bsize;
		
	for (ind = 0; need_coast_base && ind < c.nb; ind++) {	/* Loop over necessary bins only */
		
		bin = c.bins[ind];
#ifdef DEBUG
		if (debug_bin >= 0 && bin != debug_bin) continue;
#endif
		GMT_get_shore_bin (ind, &c, min_area, min_level, max_level);
		
		if (gmtdefs.verbose) fprintf (stderr, "%s: Working on block # %5d\r", GMT_program, bin);

		bin_trouble = (anti_bin == bin) ? anti_bin : -1;
			
		if (draw_coast && c.ns) {	/* Draw or dump shorelines, no need to assemble polygons */
			/* if ((np = GMT_assemble_shore (&c, direction, FALSE, shift, edge, &p)) == 0) continue; */
			if ((np = GMT_assemble_shore (&c, direction, min_level, FALSE, shift, west_border, east_border, &p)) == 0) continue;

			for (i = 0; i < np; i++) {
			  latlonlen += (p[i].n + 2)*sizeof(double);
			}
			SvGROW(lon, latlonlen);  /* pregrow for efficiency */
			SvGROW(lat, latlonlen);  /* pregrow for efficiency */

			for (i = 0; i < np; i++) {
			  sv_catpvn(lon, (char *) &GMT_d_NaN, sizeof(double)); /* separator */
			  sv_catpvn(lat, (char *) &GMT_d_NaN, sizeof(double)); /* separator */
			  for (k = 0; k < p[i].n; k++) {
                            mylon = p[i].lon[k];
			    sv_catpvn(lon, (char *) &(mylon), sizeof(double)); 
			    sv_catpvn(lat, (char *) &(p[i].lat[k]), sizeof(double)); 
			  }
			}
			
			GMT_free_polygons (p, np);
			GMT_free ((void *)p);
		}
		
		GMT_free_shore (&c);
	
	}

	if (need_coast_base) GMT_shore_cleanup (&c);

	if (gmtdefs.verbose) fprintf (stderr, "\n");
	
	if (draw_river) {	/* Read rivers file and plot as lines */
		
		if (gmtdefs.verbose) fprintf (stderr, "%s: Adding Rivers...", GMT_program);
		last_k = -1;

		for (ind = 0; ind < r.nb; ind++) {	/* Loop over necessary bins only */
		
			bin = r.bins[ind];
			GMT_get_br_bin (ind, &r, rlevels, n_rlevels);
			
			if (r.ns == 0) continue;
			
			if ((np = GMT_assemble_br (&r, shift, edge, &p)) == 0) continue;

			for (i = 0; i < np; i++) {
			  latlonlen += (p[i].n + 2)*sizeof(double);
			}
			SvGROW(lon, latlonlen);  /* pregrow for efficiency */
			SvGROW(lat, latlonlen);  /* pregrow for efficiency */

			for (i = 0; i < np; i++) {
			  sv_catpvn(lon, (char *) &GMT_d_NaN, sizeof(double)); /* separator */
			  sv_catpvn(lat, (char *) &GMT_d_NaN, sizeof(double)); /* separator */
			  for (k = 0; k < p[i].n; k++) {
                            mylon = p[i].lon[k];
			    sv_catpvn(lon, (char *) &(mylon), sizeof(double)); 
			    sv_catpvn(lat, (char *) &(p[i].lat[k]), sizeof(double)); 
			  }
			}
			
			/* Free up memory */
		
			GMT_free_br (&r);
			for (k = 0; k < np; k++) {
				GMT_free ((void *)p[k].lon);
				GMT_free ((void *)p[k].lat);
			}
			GMT_free ((void *)p);
		}
		GMT_br_cleanup (&r);
		
		if (gmtdefs.verbose) fprintf (stderr, "\n");
	}

	if (draw_border) {	/* Read borders file and plot as lines */
		
		if (gmtdefs.verbose) fprintf (stderr, "%s: Adding Borders...", GMT_program);
		
		/* Must resample borders because some points may be too far apart and look like 'jumps' */
		
		step = MAX (fabs(project_info.w - project_info.e), fabs (project_info.n - project_info.s)) / 4.0;

		last_k = -1;

		for (ind = 0; ind < b.nb; ind++) {	/* Loop over necessary bins only */
		
			bin = b.bins[ind];
			GMT_get_br_bin (ind, &b, blevels, n_blevels);
			
			if (b.ns == 0) continue;
			
			if ((np = GMT_assemble_br (&b, shift, edge, &p)) == 0) continue;

			for (i = 0; i < np; i++) {
			  latlonlen += (p[i].n + 2)*sizeof(double);
			}
			SvGROW(lon, latlonlen);  /* pregrow for efficiency */
			SvGROW(lat, latlonlen);  /* pregrow for efficiency */

			for (i = 0; i < np; i++) {
			  sv_catpvn(lon, (char *) &GMT_d_NaN, sizeof(double)); /* separator */
			  sv_catpvn(lat, (char *) &GMT_d_NaN, sizeof(double)); /* separator */
			  for (k = 0; k < p[i].n; k++) {
                            mylon = p[i].lon[k];
			    sv_catpvn(lon, (char *) &(mylon), sizeof(double)); 
			    sv_catpvn(lat, (char *) &(p[i].lat[k]), sizeof(double)); 
			  }
			}
			
			/* Free up memory */
		
			GMT_free_br (&b);
			for (k = 0; k < np; k++) {
				GMT_free ((void *)p[k].lon);
				GMT_free ((void *)p[k].lat);
			}
			GMT_free ((void *)p);
		}
		GMT_br_cleanup (&b);
		
		if (gmtdefs.verbose) fprintf (stderr, "\n");
	}
	
}

int my_GMT_begin ()
{
	/* GMT_begin will merge the command line arguments with the arguments
	 * that were used the last time this programs was called (if any).  This
	 * way one only have to give -R -J to repeat previous map projection instead
	 * of spelling out the wesn and projection parameters every time.
	 * The information is stored in the first line of the file .gmtcommands
	 * in the current directory [or optionally a provided filename] and will
	 * contain the last arguments to the common parameters like
	 * -B, -H, -J, -K, -O, -P, -R, -U, -V, -X, -Y, -c
	 * Since the meaning of -X/-Y depends on weather we have an overlay or not
	 * we maintain -X -Y for absolute shifts and -x -y for relative shifts.
	 * If the argument +file is encountered then file is used in lieu of the
	 * usual .gmtdefaults file and this argument is chopped from argv
	 */
       
	int i, j;
	char *this;

#ifdef __FreeBSD__
	/* allow divide by zero -- Inf */
	fpsetmask (fpgetmask () & ~(FP_X_DZ | FP_X_INV));
#endif 
	/* Initialize parameters */
	
	GMT_stdin  = stdin;
	GMT_stdout = stdout;

	GMT_set_home ();

	this = CNULL;
	GMT_make_fnan (GMT_f_NaN);
	GMT_make_dnan (GMT_d_NaN);
	GMT_oldargc = 0;
	frame_info.plot = FALSE;
	project_info.projection = -1;
	project_info.gave_map_width = FALSE;
	project_info.region = TRUE;
	project_info.compute_scale[0] =  project_info.compute_scale[1] = project_info.compute_scale[2] = FALSE;
	project_info.x_off_supplied = project_info.y_off_supplied = FALSE;
	project_info.region_supplied = FALSE;
	for (j = 0; j < 10; j++) project_info.pars[j] = 0.0;
	project_info.xmin = project_info.ymin = 0.0;
	project_info.z_level = DBL_MAX;	/* Will be set in map_setup */
	project_info.xyz_pos[0] = project_info.xyz_pos[1] = TRUE;
	GMT_prepare_3D ();
	gmtdefs.dlon = (project_info.e - project_info.w) / gmtdefs.n_lon_nodes;
	gmtdefs.dlat = (project_info.n - project_info.s) / gmtdefs.n_lat_nodes;
	for (i = 0; i < 4; i++) project_info.edge[i] = TRUE;
	for (i = 0; i < N_UNIQUE; i++) GMT_oldargv[i] = CNULL;
	GMT_program = "pscoast";
	GMT_grd_in_nan_value = GMT_grd_out_nan_value = GMT_d_NaN;
	
	/* Set the gmtdefault parameters from the $HOME/.gmtdefaults (if any) */
	
        /* See if user specified +optional_defaults_file */
        
	if (gmtdefs.gridfile_shorthand) GMT_setshorthand ();
	
	return;
}











